File: native-swift/TabShell/Calendar/Agenda.swift
import Foundation

/// A day the agenda lists: today, or a day with at least one event.
struct AgendaDay: Identifiable {
  let date: Date
  let events: [CalendarEvent]
  var id: Date { date }
}

struct AgendaWeek: Identifiable {
  let monthStart: Date
  let start: Date
  let days: [AgendaDay]
  var id: String { "\(monthStart.timeIntervalSince1970)-\(start.timeIntervalSince1970)" }
}

struct AgendaMonth: Identifiable {
  let start: Date
  let weeks: [AgendaWeek]
  var id: Date { start }
}

/// Groups the fixture's events by month and by week, as the agenda in `IMG_0211.PNG` does.
enum Agenda {
  static func months(events: [CalendarEvent], today: Date, calendar: Calendar = Fixture.calendar) -> [AgendaMonth] {
    var eventsByDay: [Date: [CalendarEvent]] = [calendar.startOfDay(for: today): []]
    for event in events {
      eventsByDay[calendar.startOfDay(for: event.start), default: []].append(event)
    }
    let days = eventsByDay.keys.sorted().map { date in
      AgendaDay(date: date, events: eventsByDay[date, default: []].sorted { $0.start < $1.start })
    }
    let daysByMonth = Dictionary(grouping: days) { calendar.dateInterval(of: .month, for: $0.date)!.start }
    return daysByMonth.keys.sorted().map { monthStart in
      let daysByWeek = Dictionary(grouping: daysByMonth[monthStart, default: []]) {
        calendar.dateInterval(of: .weekOfYear, for: $0.date)!.start
      }
      let weeks = daysByWeek.keys.sorted().map { weekStart in
        AgendaWeek(monthStart: monthStart, start: weekStart, days: daysByWeek[weekStart, default: []])
      }
      return AgendaMonth(start: monthStart, weeks: weeks)
    }
  }

  /// The Monday of every week from the week holding `first` to the week holding `last`.
  static func weekStarts(from first: Date, through last: Date, calendar: Calendar = Fixture.calendar) -> [Date] {
    guard var week = calendar.dateInterval(of: .weekOfYear, for: first)?.start else { return [] }
    var weeks: [Date] = []
    while week <= last {
      weeks.append(week)
      week = calendar.date(byAdding: .weekOfYear, value: 1, to: week)!
    }
    return weeks
  }

  /// "Week 38 · Sep 14 – 20"
  static func weekTitle(_ weekStart: Date, calendar: Calendar = Fixture.calendar) -> String {
    let number = calendar.component(.weekOfYear, from: weekStart)
    let end = calendar.date(byAdding: .day, value: 6, to: weekStart)!
    let range = (weekStart..<end).formatted(.interval.month(.abbreviated).day())
    return "Week \(number) · \(range)"
  }
}

File: ionic-capacitor/src/lib/agenda.ts
import type { CalendarEvent } from '../fixtures/fixture';
import { addDays, startOfDay, startOfWeek } from './dates';

/** A day the agenda lists: today, or a day with at least one event. */
export interface AgendaDay {
  date: Date;
  events: CalendarEvent[];
}

export interface AgendaWeek {
  start: Date;
  days: AgendaDay[];
}

export interface AgendaMonth {
  start: Date;
  weeks: AgendaWeek[];
}

/** Groups the events by month and by week, as the agenda in `IMG_0211.PNG` does. */
export function agendaMonths(events: CalendarEvent[], today: Date): AgendaMonth[] {
  const eventsByDay = new Map<number, CalendarEvent[]>([[startOfDay(today).getTime(), []]]);
  for (const event of events) {
    const key = startOfDay(event.start).getTime();
    eventsByDay.set(key, [...(eventsByDay.get(key) ?? []), event]);
  }

  const months: AgendaMonth[] = [];
  for (const key of [...eventsByDay.keys()].sort((a, b) => a - b)) {
    const date = new Date(key);
    const dayEvents = [...(eventsByDay.get(key) ?? [])].sort((a, b) => a.start.getTime() - b.start.getTime());

    const monthStart = new Date(date.getFullYear(), date.getMonth(), 1);
    let month = months.at(-1);
    if (!month || month.start.getTime() !== monthStart.getTime()) {
      month = { start: monthStart, weeks: [] };
      months.push(month);
    }

    const weekStart = startOfWeek(date);
    let week = month.weeks.at(-1);
    if (!week || week.start.getTime() !== weekStart.getTime()) {
      week = { start: weekStart, days: [] };
      month.weeks.push(week);
    }
    week.days.push({ date, events: dayEvents });
  }
  return months;
}

/** The Monday of every week from the week holding `first` to the week holding `last`. */
export function weekStarts(first: Date, last: Date): Date[] {
  const weeks: Date[] = [];
  for (let week = startOfWeek(first); week <= last; week = addDays(week, 7)) {
    weeks.push(week);
  }
  return weeks;
}

/** The element id of a day's first agenda row, which the week strip scrolls to. */
export function dayAnchor(date: Date): string {
  return `agenda-${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`;
}
